vms/platformvm/state/l1_validator.go

```diff
@@ -0,0 +1,439 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package state
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/google/btree"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	"github.com/ava-labs/avalanchego/utils/iterator"
+	"github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/avalanchego/utils/maybe"
+	"github.com/ava-labs/avalanchego/vms/platformvm/block"
+)
+
+var (
+	_ btree.LessFunc[L1Validator] = L1Validator.Less
+	_ utils.Sortable[L1Validator] = L1Validator{}
+
+	ErrMutatedL1Validator     = errors.New("L1 validator contains mutated constant fields")
+	ErrConflictingL1Validator = errors.New("L1 validator contains conflicting subnetID + nodeID pair")
+	ErrDuplicateL1Validator   = errors.New("L1 validator contains duplicate subnetID + nodeID pair")
+)
+
+type L1Validators interface {
+	// GetActiveL1ValidatorsIterator returns an iterator of all the
+	// active L1 validators in increasing order of EndAccumulatedFee.
+	//
+	// It is the caller's responsibility to call [Release] on the iterator after
+	// use.
+	//
+	// It is not guaranteed to be safe to modify the state while using the
+	// iterator. After releasing the iterator, the state may be safely modified.
+	GetActiveL1ValidatorsIterator() (iterator.Iterator[L1Validator], error)
+
+	// NumActiveL1Validators returns the number of currently active
+	// L1 validators.
+	NumActiveL1Validators() int
+
+	// WeightOfL1Validators returns the total active and inactive weight
+	// of L1 validators on [subnetID].
+	WeightOfL1Validators(subnetID ids.ID) (uint64, error)
+
+	// GetL1Validator returns the validator with [validationID] if it
+	// exists. If the validator does not exist, [err] will equal
+	// [database.ErrNotFound].
+	GetL1Validator(validationID ids.ID) (L1Validator, error)
+
+	// HasL1Validator returns the validator with [validationID] if it
+	// exists.
+	HasL1Validator(subnetID ids.ID, nodeID ids.NodeID) (bool, error)
+
+	// PutL1Validator inserts [l1Validator] as a validator. If the weight of the
+	// validator is 0, the validator is removed.
+	//
+	// If inserting this validator attempts to modify any of the constant fields
+	// of the L1 validator struct, an error will be returned.
+	//
+	// If inserting this validator would cause the total weight of L1
+	// validators on a subnet to overflow MaxUint64, an error will be returned.
```

Comment:
```suggestion
	// If inserting this validator would cause the total weight of L1 validators
	// on a subnet to overflow MaxUint64, an error will be returned.
```